automata/fa/fa.py

```diff
@@ -4,58 +4,30 @@
 
 import abc
 import os
-import pathlib
 import random
 import uuid
 from collections import defaultdict
-from typing import Any, Dict, Generator, List, Literal, Optional, Set, Tuple, Union
+from typing import Any, Dict, Generator, List, Optional, Set, Tuple, Union
 
 from automata.base.automaton import Automaton, AutomatonStateT
+from automata.base.utils import LayoutMethod, create_graph, save_graph
 
 # Optional imports for use with visual functionality
 try:
     import coloraide
     import pygraphviz as pgv
 except ImportError:
-    _visual_imports = False
-else:
-    _visual_imports = True
+    pass
```

Comment:
Why use a pass here? If the imports are still being used somewhere, you should still set this boolean variable and throw an exception if a visualization function gets called and these fail. It's ok to remove the pygraphviz import if that's not being used here however.

Reply:
The `pygraphviz` import is required for the type hints. `pygraphvix.AGraph` is used in this file/module. But any call to `pygraphviz` first goes through `create_graph` in `automata.base.utils` which raises an appropriate error message.
Therefore I thought of removing it here. And from the other file.

If you would want me to add it here too, I can do that. Let me know.

---

Found an interesting thing.

Even though I have `try` and `except` for the optional feature imports. I get a `NameError` if `pygraphviz` is not installed.

```bash
>>> from automata.fa.dfa import DFA
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/vipul/workspace/python/automata/automata/fa/dfa.py", line 33, in <module>
    import automata.fa.fa as fa
  File "/home/vipul/workspace/python/automata/automata/fa/fa.py", line 12, in <module>
    from automata.base.automaton import Automaton, AutomatonStateT
  File "/home/vipul/workspace/python/automata/automata/base/automaton.py", line 12, in <module>
    from automata.base.utils import freeze_value
  File "/home/vipul/workspace/python/automata/automata/base/utils.py", line 74, in <module>
    ) -> pgv.AGraph:
         ^^^
NameError: name 'pgv' is not defined
```

> Error from `utils.py`

I am thing to fix the above error by doing the following:

```python
try:
    from pygraphviz import AGraph
except ImportError:
    AGraph = None
```

Any thoughts?

---

This is should have been caught from the automated test runner. I would recommend to run the test cases with the optional dependencies and without it. To catch any such errors with may raise due to the absence of optional dependencies.

Reply:
Ahh I see. Although the solution to this issue usually comes up in the context of circular imports, I think we can do the following: https://stackoverflow.com/a/39757388/2923069

> This is should have been caught from the automated test runner. I would recommend to run the test cases with the optional dependencies and without it. To catch any such errors with may raise due to the absence of optional dependencies.

This is unfortunately a little involved. To do this, we have to set up tox to run an additional test suite that's customized to check for desired behavior if optional dependencies are not included. I'd absolutely welcome an issue / PR doing this, as it's also a prerequisite for #159, but may be a little beyond the scope of this PR right now.